src/address_space.rs

```diff
@@ -34,61 +57,156 @@ impl AddressSpace {
+        let end = start + adjusted_span;
+        for mapping in &self.mappings {
+            if mapping.addr + mapping.span < start {
+                // we have not reached the area we need yet
+                continue;
+            } else if start + adjusted_span < mapping.addr {
+                // we've moved past the area we need
+                break;
+            } else if start <= mapping.addr && mapping.addr <= start + adjusted_span {
+                // there is a mapping that starts in the space we want to place the new mapping
+                return Err("out of address space!");
+            } else if start <= mapping.addr + mapping.span
+                && mapping.addr + mapping.span <= start + adjusted_span
+            {
+                // there is a mapping that ends in the space we want to place the new mapping
+                return Err("out of address space!");
+            } else if mapping.addr < start && start + adjusted_span < mapping.addr + mapping.span {
+                // there is a mapping covers the space we want to place the new mapping
+                return Err("out of address space!");
+            }
+        }
+        // if we have enough space for our new mapping
+        let new_mapping = MapEntry::new(source.clone(), offset, adjusted_span, start, flags);
+        self.mappings.push(new_mapping);
+        self.mappings.sort_by(|a, b| a.addr.cmp(&b.addr));
+        return Ok(());
     }
 
     /// Remove the mapping to `DataSource` that starts at the given address.
     ///
     /// # Errors
     /// If the mapping could not be removed.
     pub fn remove_mapping<D: DataSource>(
-        &self,
-        source: &D,
+        &mut self,
+        source: Arc<D>,
         start: VirtualAddress,
     ) -> Result<(), &str> {
-        todo!()
+        let mapping = self.get_mapping_for_addr(start).unwrap();
+        let s = &self.mappings.len();
+
+        if let Some(pos) = self.mappings.iter().position(|x| x.addr == mapping.addr) {
+            self.mappings.remove(pos);
+        } else {
+            return Err("error removing mapping");
+        }
+        if self.mappings.len() != s - 1 {
+            return Err("error removing mapping");
+        }
+        drop(source);
+        return Ok(());
     }
 
     /// Look up the DataSource and offset within that DataSource for a
     /// VirtualAddress / AccessType in this AddressSpace
-    /// 
+    ///
     /// # Errors
     /// If this VirtualAddress does not have a valid mapping in &self,
     /// or if this AccessType is not permitted by the mapping
+    /// access is read,write,and execute
+    /// private vs cow vs shared is about cache sharing
     pub fn get_source_for_addr<D: DataSource>(
         &self,
         addr: VirtualAddress,
-        access_type: FlagBuilder
-    ) -> Result<(&D, usize), &str> {
-        todo!();
+        access_type: FlagBuilder,
+    ) -> Result<(Arc<dyn DataSource>, usize), &str> {
+        if !access_type.read {
```

Comment:
I'm a little bit confused about this line. `access_type` is declaring what the caller wants to access, right? So you should still need to check it against some internal representation of their permission.

Also, even if they don't have read permissions, would they still not be able to write, in principle?